KEP-4471: kubeadm: make a control-plane's kubelet talk to the local API Server on kubeadm join. (#4496)
kubernetes/enhancements: keps/sig-cluster-lifecycle/kubeadm/4471-cp-join-kubelet-local-apiserver/README.md

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As a kubeadm user, I wish no component violates the kubernetes version skew policy when joining a control plane node.

#### Story 2

As a kubeadm user, I wish the kubelet of a joining control plane node points to the local kube-apiserver.

#### Story 3

As a kubeadm user, I wish the GA API of kubeadm to be stable and breaking changes to it to be announced ahead of time.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- This change is only relevant for the control plane nodes because worker nodes do
  not have a local kube-apiserver to point to and are not affected by the potential

>   version skew policy violation.

Author: Tried to clarify that point :-)

for the kubelet's bootstrap to complete into a separate phase named `KubeletWaitBootstrapPhase`
gets extracted.

1. `PreflightPhase`
2. `ControlPlanePreparePhase`
3. `CheckEtcdPhase`
4. `KubeletStartPhase`
5. **`KubeletWaitBootstrapPhase`**
6. `ControlPlaneJoinPhase`

Second the `EtcdLocalSubphase` which is the first part of the `ControlPlanJoinPhase` gets moved
to a new phase `ControlPlaneJoinEtcdPhase` and added between the `KubeletStartPhase` and
`KubeletWaitBootstrapPhase`.

1. `PreflightPhase`
2. `ControlPlanePreparePhase`
3. `CheckEtcdPhase`
4. `KubeletStartPhase`
5. **`ControlPlaneJoinEtcdPhase`**
6. `KubeletWaitBootstrapPhase`
7. `ControlPlaneJoinPhase`

Actions to preserve the old behavior when the feature gate is disabled:

- There should be no duplication of code, instead the relevant shared code parts should be moved
  to private functions so both phases could execute them, depending on the feature gate.
- The new phases very lightweight and make use of the private functions.
- The phases `KubeletStartPhase` and `ControlPlaneJoinPhase` should behave as before and run the above introduced private functions.

> - Nothing should be done during the new phases `ControlPlaneJoinEtcdPhase` and `KubeletWaitBootstrapPhase`.

Reviewer: one thing we can do is to add "EXPERIMENTAL: some phase description" in the descriptions for all new phases until the FG graduates to GA.

- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature gate `ControlPlaneKubeletLocalMode`
- Initial unit and e2e tests completed and enabled
- [Document the feature gate](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#feature-gates).

> - Document the upcoming breaking change at the release notes.

Reviewer: ```suggestion
- Document the upcoming breaking change at the release notes
```

or switch all of them to have .

#### Beta

- Make feature gate to be enabled by default
- Gather feedback from developers and surveys
- Make unit and e2e test changes
- Update the feature gate documentation
- [Document the new phases](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join-phase/)

#### GA

- Gather feedback from developers and surveys
- Update unit tests
- Remove e2e tests as this will be tested by all existing kubeadm e2e tests
- Update the feature gate documentation
- Update the phases documentation

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

> - N/A -> Alpha: users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to before calling `kubeadm upgrade apply` This will allow them to enable the `ControlPlaneKubeletLocalMode` in case they wish to join control plane nodes to this cluster. This scenario is anticipated as rare, because usually users maintain a stable control plane with 3 or more members before upgrading it. But it is still plausible and can be documented in the feature gate documentation.

Reviewer: ```suggestion
- N/A -> Alpha: users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to before enable the `ControlPlaneKubeletLocalMode` feature gate, before calling `kubeadm upgrade apply`. This will allow them to join control plane nodes with this feature enabled. This scenario is anticipated as rare, because usually users maintain a stable control plane with 3 or more members before upgrading it. But it is still plausible and can be documented in the feature gate documentation.
```

we can add another paragraph here, explaining that users who wish to manually migrate existing nodes to use the new way can just edit their "kubelet.conf" files on CP nodes? the server there would match the one we already have for kube-scheduler and kube-controller-manager kubeconfig files.

- Alpha -> Beta: similarly to the previous stage users can modify the `ClusterConfiguration` to disable the feature gate during upgrade. This will allow them to use the "old way", in case they wish to join more control plane nodes to the cluster while the feature gate is enabled by default.

> - Beta -> GA: users could no longer patch the `ClusterConfiguration` to opt-out of the feature and it will be locked to default.

Reviewer: ```suggestion
- Beta -> GA: users could no longer patch the `ClusterConfiguration` to opt-out of the feature and it will be locked to enabled by default.
```

- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

Not applicable.

## Production Readiness Review Questionnaire

kubeadm is considered an "out of tree" component and PRR is out of scope.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

- 01.02.2024: KEP issue created.
- 08.02.2024: KEP draft created.

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> Moving or introducing new phases are breaking changes to the API of kubeadm.

Reviewer: ```suggestion
Moving or introducing new phases are breaking changes to the CLI of kubeadm.
```

Adding new phases may break users which are executing the single phases manually.
The changes must be well-documented in release notes for users to adapt the changes.

The reordering of phases may change delays or timeouts when control plane nodes are joined.
Modifying this code path may introduce potential for user complains about HA cluster creation and maintenance with kubeadm.
Sufficient testing and gathering feedback from users would be mandatory.

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

### Take the risk of violating kubernetes version skew policy

> By not introducing the change, users of kubeadm are on risk to hit a case in future where a violation of the kubernetes version skew policy happens and joining a control plane node fails.

Reviewer: ```suggestion
By not introducing the change, users of kubeadm are on risk to hit a case in the future where a violation of the Kubernetes version skew policy happens and joining a control plane node may fail.
```

> However such a case is rare.

Reviewer: ```suggestion
We anticipate such cases to be rare.
```

enhancement:
- Does this enhancement involve coordinating behavior in the control plane and nodes?
- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

Not applicable.

## Production Readiness Review Questionnaire

kubeadm is considered an "out of tree" component and PRR is out of scope.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

- 01.02.2024: KEP issue created.

> - 08.02.2024: KEP draft created.

Reviewer: ```suggestion
- 08.02.2024: KEP draft created.
- <date-of-change>: KEP marked as implementable for 1.31 or later
```
